services/collaboration/pkg/connector/fileconnector.go

```diff
@@ -472,6 +525,482 @@ func (f *FileConnector) UnLock(ctx context.Context, lockID string) (string, erro
+// name. This new name will be generated by prefixing a random string to the
+// suggested name.
+// Since the upload won't use any lock, the upload will fail if the target file
+// already exists and it isn't empty. This means that, this method can only
+// generate new files.
+func (f *FileConnector) PutRelativeFileSuggested(ctx context.Context, ccs ContentConnectorService, stream io.Reader, streamLength int64, target string) (*PutRelativeResponse, error) {
+	// assume the target is a full name
+	wopiContext, err := middleware.WopiContextFromCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	logger := zerolog.Ctx(ctx).With().
+		Str("PutTarget", target).
+		Logger()
+
+	// stat the current file in order to get the reference of the parent folder
+	oldStatRes, err := f.gwc.Stat(ctx, &providerv1beta1.StatRequest{
+		Ref: wopiContext.FileReference,
+	})
+	if err != nil {
+		logger.Error().Err(err).Msg("PutRelativeFileSuggested: stat failed")
+		return nil, err
+	}
+
+	if oldStatRes.GetStatus().GetCode() != rpcv1beta1.Code_CODE_OK {
+		logger.Error().
+			Str("StatusCode", oldStatRes.GetStatus().GetCode().String()).
+			Str("StatusMsg", oldStatRes.GetStatus().GetMessage()).
+			Msg("PutRelativeFileSuggested: stat failed with unexpected status")
+		return nil, NewConnectorError(500, oldStatRes.GetStatus().GetCode().String()+" "+oldStatRes.GetStatus().GetMessage())
+	}
+
+	if strings.HasPrefix(target, ".") {
+		// the target is an extension, so we need to use the original
+		// name with the modified extension
+		oldStatPath := oldStatRes.GetInfo().GetPath()
+		ext := path.Ext(oldStatPath)
+		target = strings.TrimSuffix(path.Base(oldStatPath), ext) + target
+	}
+
+	finalTarget := target
+	newLogger := logger
+	for isDone := false; !isDone; {
+		var conError *ConnectorError
+
+		targetPath := utils.MakeRelativePath(finalTarget)
+		// need to change the file reference of the wopicontext to point to the new path
+		wopiContext.FileReference = &providerv1beta1.Reference{
+			ResourceId: oldStatRes.GetInfo().GetParentId(),
+			Path:       targetPath,
+		}
+
+		// create a new context for the modified wopicontext
+		newLogger := logger.With().Str("NewFileReference", wopiContext.FileReference.String()).Logger()
+		newCtx := middleware.WopiContextToCtx(newLogger.WithContext(ctx), wopiContext)
+
+		// try to put the file. It mustn't return a 400 or 409
+		_, err := ccs.PutFile(newCtx, stream, streamLength, "")
+		if err != nil {
+			// if the error isn't a connectorError, fail the request
+			if !errors.As(err, &conError) {
+				newLogger.Error().Err(err).Msg("PutRelativeFileSuggested: put file failed")
+				return nil, err
+			}
+
+			if conError.HttpCodeOut == 409 {
+				// if conflict generate a different name and retry.
+				// this should happen only once
+				actualFilename, _ := f.extractFilenameAndPrefix(target)
+				finalTarget = f.generatePrefix() + " " + actualFilename
+			} else {
+				// TODO: code 400 might happen, what to do?
```

Comment:
Why not check for it?

Reply:
On the one hand, the `PutFile` method of the content connector doesn't return that code, so we probably can skip the check, at least for now.
On the other hand, the WOPI spec says that the `PutRelative` with the `suggestedTarget` header (which goes through this method) MUST NOT return a 400 code, so we might want to ensure it will never happen. In this case, the expectation is that we use a different name for the file. However, if we assume we get a 400 because of a bad file name (as said above, this isn't happening right now), how can we generate a different file name based on the illegal name? Note that we want to ensure somehow that the new file name is valid and won't return a 400 error.

The code should be fine, so we can skip the check. It will forward the error code from the `PutFile` requests, which shouldn't be a 400.